### PR TITLE
Add HackMD to incubator

### PR DESCRIPTION
**Pull request by yuanying**
*****************************************************************************************

Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the 
history. This will make it easier to identify new changes. The PR will be squashed 
anyways when it is merged. Thanks.

*****************************************************************************************

### CODE REVIEW

File: incubator/hackmd/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            - name: HMD_DB_URL
+              value: postgres://{{ .Values.postgresql.postgresUser }}:{{ .Values.postgresql.postgresPassword }}@{{ template "postgresql.fullname" . }}:5432/{{ .Values.postgresql.postgresDatabase }}
+            {{- range $key, $value :=  .Values.env }}
+            - name: {{ $key | upper | replace "-" "_" }}
+              value: {{ $value | quote }}
+            {{- end }}
+          volumeMounts:
+            - name: config
+              subPath: config.json
+              mountPath: /hackmd/config.json
```
Comment (SISheogorath) [inline incubator/hackmd/templates/deployment.yaml]: You actually don't need this if you use environment variables to configure HackMD.

https://github.com/hackmdio/hackmd#environment-variables-will-overwrite-other-server-configs

File: incubator/hackmd/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            - name: config
+              subPath: .sequelizerc
+              mountPath: /hackmd/.sequelizerc
```
Comment (SISheogorath) [inline incubator/hackmd/templates/deployment.yaml]: This file is build to work with environment variables for configuration.

**Review by SISheogorath — COMMENTED** [review]
Just some hints to improve the container handling and less persistent data :)

**Comment by yuanying** [comment]
@SISheogorath Hi, thanks for your suggestion! I fixed it and works well!

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the PR. I'd like you to make a few changes.

* Please compare with what `helm create` using Helm 2.8+ gives you and update accordingly.
* Please follow our best practice for storage class configuration. See #1869.
* By default, Postgres auto-generates a password. Keep it that way.

**Comment by SISheogorath** [comment]
Just as question, when postgres autogenerates the password, how is it passed to HackMD?

**Comment by unguiculus** [comment]
@SISheogorath You already have that correct.

**Comment by SISheogorath** [comment]
This indicates that the postgres user for HackMD doesn't exist. I guess based on the next lines in the log that it's creating the role afterwards, which means HackMD is started and able to connect to the postgres instance before the users are setup.

The HackMD docker container is build to try to connect to the database, using jwilder's dockerize, to make sure that it's really available and ready (which doesn't have to be the case, when the users are created afterwards). Then it starts the actual HackMD instance.

https://github.com/hackmdio/docker-hackmd/blob/9fdc72414a1a1afd32d6f558bea6697a4197ce75/docker-entrypoint.sh#L10

So while in theory when the order is correct, everything works fine, I have no idea how to handle a race condition that make us checking for an non-existing user. The DB shouldn't be available before it's set up properly.

I can't say much about the way the K8s template handles this but I know the default docker image works well with our setups. Maybe someone who already uses postgres in K8s can tell something about the same or similar problems and their solution :)

File: incubator/hackmd/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+              port: 3000
+            initialDelaySeconds: 30
+          env:
+            - name: HMD_DB_PASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: {{ template "hackmd.postgresql.fullname" . }}
+                  key: postgres-password
+            - name: HMD_DB_URL
+              value: postgres://{{ .Values.postgresql.postgresUser }}:$(HMD_DB_PASSWORD)@{{ template "hackmd.postgresql.fullname" . }}:5432/{{ .Values.postgresql.postgresDatabase }}
```
Comment (yuanying) [inline incubator/hackmd/templates/deployment.yaml]: At previous post, I mistook the DB_URL (I was using {{ template "hacked.fullname" . }} as DB host). Maybe it caused this problem?

**Review by yuanying — COMMENTED** [review]
Anyway I have changed my codes as you have indicated.

File: incubator/hackmd/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+appVersion: "1.0"
+description: A Helm chart for Kubernetes
+name: hackmd
+version: 0.1.0
```
Comment (unguiculus) [inline incubator/hackmd/Chart.yaml]: Please add more info, such as home, sources, maintainers, keywords, icon.

File: incubator/hackmd/templates/_helpers.tpl [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+{{- printf "%s-%s" .Chart.Name .Chart.Version | replace "+" "_" | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{/*
+Create a default fully qualified postgresql name.
+We truncate at 24 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+*/}}
+{{- define "hackmd.postgresql.fullname" -}}
+{{- $name := default "postgresql" .Values.postgresql.nameOverride -}}
+{{- printf "%s-%s" .Release.Name $name | trunc 24 | trimSuffix "-" -}}
```
Comment (unguiculus) [inline incubator/hackmd/templates/_helpers.tpl]: 24 -> 63

File: incubator/hackmd/values.yaml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+## Configuration values for the postgresql dependency.
+## ref: https://github.com/kubernetes/charts/blob/master/stable/postgresql/README.md
+##
+postgresql:
+  install: true
+  imageTag: "9.6.2"
+
+  postgresUser: hackmd
+  postgresDatabase: hackmd
```
Comment (unguiculus) [inline incubator/hackmd/values.yaml]: Don't set a default. By default, the password is auto-generated. Let's keep it that way.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/hackmd/Chart.yaml (hunk 4 above)
Comment (yuanying) [reply incubator/hackmd/Chart.yaml]: Done

File: incubator/hackmd/templates/_helpers.tpl (hunk 5 above)
Comment (yuanying) [reply incubator/hackmd/templates/_helpers.tpl]: Done

File: incubator/hackmd/values.yaml (hunk 6 above)
Comment (yuanying) [reply incubator/hackmd/values.yaml]: Should I also remove username and database name?

**Comment by yuanying** [comment]
@unguiculus Could you check this?

File: incubator/hackmd/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+- markdown
+- http
+- web
+- application
+home: https://hackmd.io
+sources:
+- https://github.com/hackmdio/hackmd
+- https://github.com/hackmdio/docker-hackmd
+maintainers:
+- name: OTSUKA, Yuanying
```
Comment (unguiculus) [inline incubator/hackmd/Chart.yaml]: Use Github username

Comment (unguiculus) [inline incubator/hackmd/Chart.yaml]: Use Github username

Comment (unguiculus) [inline incubator/hackmd/Chart.yaml]: Use Github username

Comment (yuanying) [reply incubator/hackmd/Chart.yaml]: Done.

File: incubator/hackmd/values.yaml (hunk 6 above)
Comment (unguiculus) [reply incubator/hackmd/values.yaml]: Just the password would be ok.

File: incubator/hackmd/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+              valueFrom:
+                secretKeyRef:
+                  name: {{ template "hackmd.postgresql.fullname" . }}
+                  key: postgres-password
+              {{- else }}
+              value: {{ .Values.postgresql.postgresPassword }}
+              {{- end }}
+            - name: HMD_DB_URL
+              value: postgres://{{ .Values.postgresql.postgresUser }}:$(HMD_DB_PASSWORD)@{{ template "hackmd.database.host" . }}:5432/{{ .Values.postgresql.postgresDatabase }}
+            {{- range $key, $value :=  .Values.env }}
```
Comment (unguiculus) [inline incubator/hackmd/templates/deployment.yaml]: Please rename it to `extraVars` and follow this more flexible style:

```yaml
        {{- if .Values.extraVars }}
          env:
{{ toYaml .Values.extraVars | indent 12 }}
        {{- end }}
```

File: incubator/hackmd/values.yaml [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  ## database data Persistent Volume Storage Class
+  ## If defined, storageClassName: <storageClass>
+  ## If set to "-", storageClassName: "", which disables dynamic provisioning
+  ## If undefined (the default) or set to null, no storageClassName spec is
+  ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
+  ##   GKE, AWS & OpenStack)
+  ##
+  # storageClass: "-"
+
+env: {}
```
Comment (unguiculus) [inline incubator/hackmd/values.yaml]: With changes suggested above, this should be a list.

`env: []`

File: incubator/hackmd/values.yaml [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+affinity: {}
+
+## Configuration values for the postgresql dependency.
+## ref: https://github.com/kubernetes/charts/blob/master/stable/postgresql/README.md
+##
+postgresql:
+  install: true
+  imageTag: "9.6.2"
+  postgresUser: "hackmd"
+  postgresDatabase: "hackmd"
```
Comment (unguiculus) [inline incubator/hackmd/values.yaml]: Leave password empty so it is auto-generated by default.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/hackmd/templates/deployment.yaml (hunk 8 above)
Comment (yuanying) [reply incubator/hackmd/templates/deployment.yaml]: Done

File: incubator/hackmd/values.yaml (hunk 9 above)
Comment (yuanying) [reply incubator/hackmd/values.yaml]: Done

File: incubator/hackmd/values.yaml (hunk 10 above)
Comment (yuanying) [reply incubator/hackmd/values.yaml]: Yes, I don't specified postgresPassword here.

**Comment by unguiculus** [comment]
Next time, please don't squash commits so I can see what you changed.

File: incubator/hackmd/values.yaml (hunk 10 above)
Comment (unguiculus) [reply incubator/hackmd/values.yaml]: Ah, sorry, you're right.

**Comment by unguiculus** [comment]
Now, please move the chart directly to stable.

**Comment by yuanying** [comment]
@unguiculus ok, thanks for your reviewing!

**Review by unguiculus — APPROVED** [review]

**Comment by SISheogorath** [comment]
Thanks a lot for your work! I'll add a guide to the HackMD readme.

**Comment by dcwangmit01** [comment]
This PR looks to be a copy of the code from:
https://github.com/hackmdio/docker-hackmd
Which is a fork from:
https://github.com/a-r-g-v/docker-hackmd

The latter does not mention any license.  The former links to the license in the hackmd sources with changed from MIT to AGPLv3 in November 2017.  We should probably ask the owners of both repo's mentioned above to clarify their licenses to be copyleft.  Kubernetes is Apache 2.0.

**Comment by yuanying** [comment]
@dcwangmit01 These code are actually copied from:
* https://github.com/yuanying/docker-hackmd/commit/ba16ddd01dd8b0a37b84b804cec9dc51ca543217

So, I have copyright of these code I think.
If there are any problem of this PR, Could you elaborate about that?

**Pull request merged**